zebra-chain/src/value_balance/tests/prop.rs

```diff
@@ -88,4 +88,14 @@ proptest! {
             _ => prop_assert!(matches!(collection.iter().sum(), Err(ValueBalanceError::AmountError(_))))
         }
     }
+
+    #[test]
+    fn value_balance_serialization(value_balance in any::<ValueBalance<NegativeAllowed>>()) {
```

Comment:
We'll also need a "bytes to ValueBalance to bytes" round-trip proptest.
If deserialization errors, just skip serialization.

Can you make sure that `Amount` has struct and bytes proptests as well?

Reply:
As we have it now the deserialization will never fail even if the arbitrary bytes are bigger or smaller than the allowed range. I added a test for this in https://github.com/ZcashFoundation/zebra/pull/2554/commits/f18499cd33f4a4f7574d1b8c3096d0254b763fa3

I am unsure if we need to change the behaviour, probably we do.

Reply:
Yeah I think you're right.

We need `Amount::from_bytes` and `ValueBalance::from_bytes` to fail if the amount is outside the constraint.

But they are only storing valid data that Zebra has committed to the database. So it's ok for them to call `try_into().expect(...)`, as long as we document that they panic in the method docs.

Reply:
I am now returning the errors from the `from_bytes()` methods: https://github.com/ZcashFoundation/zebra/pull/2554/commits/e89760393c801007a470d3e1e30ebe135fdc9887

One issue i see is that that the proptests generates bytes so big that the block of code inside of `let Ok() ...` will kind of never be executed.

Reply:
> One issue i see is that that the proptests generates bytes so big that the block of code inside of `let Ok() ...` will kind of never be executed.

Proptests are also supposed to test with some fixed values, and I think all-zeroes is one of them. But we shouldn't rely on that.

Can you add struct serialization and bytes deserialization proptests for `Amount` as well?
We'll get better coverage that way, because around 1 in 4000 random byte values are a valid amount.

Reply:
https://github.com/ZcashFoundation/zebra/pull/2554/commits/92e4e6c9ff0ed0a69ef82941fa29e4b075341179